lib/Pulse.Lib.AVLTree.fsti

```diff
@@ -0,0 +1,110 @@
+//Pulse AVL tree interface. Inspired from Steel. The FStar spec file is adopted from Steel
+//----------------------------------------------------------------------------------------------------------
+module Pulse.Lib.AVLTree
+open Pulse.Lib.Pervasives
+module S = Pulse.Lib.Stick.Util
+module FA = Pulse.Lib.Forall.Util
+open FStar.List.Tot
+
+module T = Spec.AVLTree
+module G = FStar.Ghost
+
+noeq
+type node (t:Type0) = {
+    data : t;
+    left : tree_t t;
+    right : tree_t t;
+}
+
+and node_ptr (t:Type0) = ref (node t)
+
+//A nullable pointer to a node
+and tree_t (t:Type0) = option (node_ptr t)
+
+let vprop = slprop
+
+val is_tree #t (ct:tree_t t) (ft:T.tree t)
+: Tot vprop (decreases ft)
+
+val height (#t:Type0) (x:tree_t t) (#ft:G.erased (T.tree t))
+  : stt nat
+(requires is_tree x ft)
+(ensures fun n -> is_tree x ft ** pure (n == T.height ft))
+
+val is_empty (#t:Type) (x:tree_t t) (#ft:(T.tree t))
+   : stt bool
+  (requires is_tree x ft)
+  (ensures fun b -> is_tree x ft ** pure (b <==> (ft == T.Leaf)))
+
+val null_tree_t (t:Type0) : tree_t t
+
+val create (t:Type0)
+   : stt (tree_t t)
+  (requires emp)
+  (ensures fun x -> is_tree x T.Leaf)
+
+val append_left_none (#t:Type0) (x:tree_t t) (v:t) (#ft:G.erased (T.tree t))
```

Comment:
Do we need the appends, rotates, balalance etc. functions in the interface? Can't the interface just have create, insert, delete, mem, and height?

Reply:
comment addressed.